File: android/app/src/main/kotlin/com/splitfree/core/Money.kt
package com.splitfree.core

import java.math.BigDecimal
import java.math.RoundingMode
import java.text.NumberFormat
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap

/**
 * An exact monetary amount stored as integer minor units (cents, paise, yen, …).
 *
 * Every amount in SplitFree is an integer. Splitting, settling and balance math
 * all operate on minor units so a bill can never gain or lose a cent to binary
 * floating-point drift. [BigDecimal] and [Double] appear only at the edges:
 * parsing what a person typed, and formatting what they read back.
 *
 * This mirrors `Money` in the iOS app exactly. The two must agree, or the same
 * expense would produce different splits on different phones.
 */
data class Money(val minorUnits: Long, val currencyCode: String) {

    val isZero: Boolean get() = minorUnits == 0L
    val isNegative: Boolean get() = minorUnits < 0
    val isPositive: Boolean get() = minorUnits > 0
    val magnitude: Money get() = Money(kotlin.math.abs(minorUnits), currencyCode)

    /** The amount in major units, for display and rate math only. */
    val decimalValue: BigDecimal
        get() = BigDecimal(minorUnits).movePointLeft(Currencies.fractionDigits(currencyCode))

    operator fun plus(other: Money): Money {
        require(currencyCode == other.currencyCode) { "Cannot add $currencyCode to ${other.currencyCode}" }
        return Money(minorUnits + other.minorUnits, currencyCode)
    }

    operator fun minus(other: Money): Money {
        require(currencyCode == other.currencyCode) { "Cannot subtract ${other.currencyCode} from $currencyCode" }
        return Money(minorUnits - other.minorUnits, currencyCode)
    }

    operator fun unaryMinus(): Money = Money(-minorUnits, currencyCode)

    /** Formats in the user's locale but this amount's own currency. */
    fun formatted(showsSign: Boolean = false, showsCurrencyCode: Boolean = false): String =
        CurrencyFormatting.string(minorUnits, currencyCode, showsSign, showsCurrencyCode)

    /**
     * Magnitude only, for places where the surrounding copy carries the sign
     * ("you owe" / "owes you") rather than a minus glyph.
     */
    fun formattedAbsolute(showsCurrencyCode: Boolean = false): String =
        magnitude.formatted(showsCurrencyCode = showsCurrencyCode)

    companion object {
        fun zero(currencyCode: String) = Money(0, currencyCode)

        fun fromDecimal(amount: BigDecimal, currencyCode: String): Money {
            val scale = Currencies.fractionDigits(currencyCode)
            // HALF_EVEN matches the banker's rounding the iOS app uses.
            val scaled = amount.movePointRight(scale).setScale(0, RoundingMode.HALF_EVEN)
            return Money(scaled.toLong(), currencyCode)
        }
    }
}

object CurrencyFormatting {

    // NumberFormat construction is expensive and these are reused constantly.
    private val cache = ConcurrentHashMap<String, NumberFormat>()

    fun string(
        minorUnits: Long,
        currencyCode: String,
        showsSign: Boolean = false,
        showsCurrencyCode: Boolean = false,
    ): String {
        val formatter = formatter(currencyCode, showsCurrencyCode)
        val digits = Currencies.fractionDigits(currencyCode)
        val value = BigDecimal(minorUnits).movePointLeft(digits)
        val base = formatter.format(value)
        return if (showsSign && minorUnits > 0) "+$base" else base
    }

    private fun formatter(currencyCode: String, showsCurrencyCode: Boolean): NumberFormat =
        cache.getOrPut("$currencyCode|$showsCurrencyCode") {
            val digits = Currencies.fractionDigits(currencyCode)
            val format = NumberFormat.getCurrencyInstance(Locale.getDefault())
            runCatching {
                format.currency = java.util.Currency.getInstance(currencyCode)
            }
            format.minimumFractionDigits = digits
            format.maximumFractionDigits = digits
            if (format is java.text.DecimalFormat) {
                val symbols = format.decimalFormatSymbols
                // Java only knows the symbols for currencies it ships; fall back
                // to our own table so every one of the 154 renders sensibly.
                symbols.currencySymbol =
                    if (showsCurrencyCode) currencyCode.uppercase() else Currencies.symbol(currencyCode)
                format.decimalFormatSymbols = symbols
            }
            format
        }

    /** Parses free-typed input ("12.50", "1 234,56", "$8") into minor units. */
    fun parse(text: String, currencyCode: String): Long? {
        val stripped = text.filter { it.isDigit() || it == '.' || it == ',' || it == '-' }
        if (stripped.isEmpty()) return null

        // Whichever separator comes last is the decimal separator; the other groups.
        val lastDot = stripped.lastIndexOf('.')
        val lastComma = stripped.lastIndexOf(',')
        val normalized = when {
            lastDot >= 0 && lastComma >= 0 ->
                if (lastDot > lastComma) stripped.replace(",", "")
                else stripped.replace(".", "").replace(',', '.')
            lastComma >= 0 -> stripped.replace(',', '.')
            else -> stripped
        }

        val decimal = normalized.toBigDecimalOrNull() ?: return null
        return Money.fromDecimal(decimal, currencyCode).minorUnits
    }

    fun symbol(currencyCode: String): String = Currencies.symbol(currencyCode)
}

File: SplitFree/Core/Money.swift
import Foundation

/// An exact monetary amount stored as integer minor units (cents, paise, yen, …).
///
/// Every amount in SplitFree is an integer. Splitting, settling and balance math
/// all operate on minor units so a bill can never gain or lose a cent to binary
/// floating-point drift. `Decimal`/`Double` appear only at the edges: parsing what
/// a person typed, and formatting what they read back.
struct Money: Hashable, Codable, Sendable {
    /// Amount in the currency's smallest unit. Signed: negative means "owed to".
    var minorUnits: Int
    /// ISO 4217 alphabetic code, e.g. `USD`.
    var currencyCode: String

    init(minorUnits: Int, currencyCode: String) {
        self.minorUnits = minorUnits
        self.currencyCode = currencyCode
    }

    init(amount: Decimal, currencyCode: String) {
        let scale = Currency.fractionDigits(for: currencyCode)
        let factor = pow(Decimal(10), scale)
        var raw = amount * factor
        var rounded = Decimal()
        NSDecimalRound(&rounded, &raw, 0, .bankers)
        self.minorUnits = NSDecimalNumber(decimal: rounded).intValue
        self.currencyCode = currencyCode
    }

    static func zero(_ currencyCode: String) -> Money {
        Money(minorUnits: 0, currencyCode: currencyCode)
    }

    var isZero: Bool { minorUnits == 0 }
    var isNegative: Bool { minorUnits < 0 }
    var isPositive: Bool { minorUnits > 0 }
    var magnitude: Money { Money(minorUnits: abs(minorUnits), currencyCode: currencyCode) }

    /// The amount as a decimal in major units - for display and for rate math only.
    var decimalValue: Decimal {
        let scale = Currency.fractionDigits(for: currencyCode)
        return Decimal(minorUnits) / pow(Decimal(10), scale)
    }

    var doubleValue: Double { NSDecimalNumber(decimal: decimalValue).doubleValue }

    static func + (lhs: Money, rhs: Money) -> Money {
        precondition(lhs.currencyCode == rhs.currencyCode, "Cannot add \(lhs.currencyCode) to \(rhs.currencyCode)")
        return Money(minorUnits: lhs.minorUnits + rhs.minorUnits, currencyCode: lhs.currencyCode)
    }

    static func - (lhs: Money, rhs: Money) -> Money {
        precondition(lhs.currencyCode == rhs.currencyCode, "Cannot subtract \(rhs.currencyCode) from \(lhs.currencyCode)")
        return Money(minorUnits: lhs.minorUnits - rhs.minorUnits, currencyCode: lhs.currencyCode)
    }

    static prefix func - (value: Money) -> Money {
        Money(minorUnits: -value.minorUnits, currencyCode: value.currencyCode)
    }

    static func += (lhs: inout Money, rhs: Money) { lhs = lhs + rhs }
    static func -= (lhs: inout Money, rhs: Money) { lhs = lhs - rhs }

    static func < (lhs: Money, rhs: Money) -> Bool { lhs.minorUnits < rhs.minorUnits }
}

extension Money {
    /// Formats using the user's locale but the amount's own currency.
    func formatted(showsSign: Bool = false, showsCurrencyCode: Bool = false) -> String {
        CurrencyFormatting.string(
            minorUnits: minorUnits,
            currencyCode: currencyCode,
            showsSign: showsSign,
            showsCurrencyCode: showsCurrencyCode
        )
    }

    /// Magnitude only - used where the sign is carried by surrounding copy
    /// ("you owe" / "owes you") rather than a minus glyph.
    func formattedAbsolute(showsCurrencyCode: Bool = false) -> String {
        magnitude.formatted(showsCurrencyCode: showsCurrencyCode)
    }
}

enum CurrencyFormatting {
    private static let cache = FormatterCache()

    static func string(
        minorUnits: Int,
        currencyCode: String,
        showsSign: Bool = false,
        showsCurrencyCode: Bool = false
    ) -> String {
        let formatter = cache.formatter(for: currencyCode, showsCurrencyCode: showsCurrencyCode)
        let scale = Currency.fractionDigits(for: currencyCode)
        let value = Decimal(minorUnits) / pow(Decimal(10), scale)
        let base = formatter.string(from: NSDecimalNumber(decimal: value)) ?? "\(value)"
        if showsSign && minorUnits > 0 { return "+" + base }
        return base
    }

    /// Parses free-typed input ("12.50", "1 234,56", "$8") into minor units.
    static func parse(_ text: String, currencyCode: String) -> Int? {
        let stripped = text.filter { $0.isNumber || $0 == "." || $0 == "," || $0 == "-" }
        guard !stripped.isEmpty else { return nil }

        // Whichever separator appears last is the decimal separator; the other groups.
        let lastDot = stripped.lastIndex(of: ".")
        let lastComma = stripped.lastIndex(of: ",")
        var normalized = stripped
        switch (lastDot, lastComma) {
        case let (dot?, comma?):
            if dot > comma {
                normalized = stripped.replacingOccurrences(of: ",", with: "")
            } else {
                normalized = stripped.replacingOccurrences(of: ".", with: "")
                normalized = normalized.replacingOccurrences(of: ",", with: ".")
            }
        case (nil, .some):
            normalized = stripped.replacingOccurrences(of: ",", with: ".")
        default:
            break
        }

        guard let decimal = Decimal(string: normalized, locale: Locale(identifier: "en_US_POSIX")) else { return nil }
        return Money(amount: decimal, currencyCode: currencyCode).minorUnits
    }

    /// A short symbol suitable for inline use next to a text field.
    static func symbol(for currencyCode: String) -> String {
        Currency.symbol(for: currencyCode)
    }
}

/// `NumberFormatter` construction is expensive; these are reused across the app.
private final class FormatterCache: @unchecked Sendable {
    private var storage: [String: NumberFormatter] = [:]
    private let lock = NSLock()

    func formatter(for currencyCode: String, showsCurrencyCode: Bool) -> NumberFormatter {
        let key = "\(currencyCode)|\(showsCurrencyCode)"
        lock.lock()
        defer { lock.unlock() }
        if let existing = storage[key] { return existing }
        let formatter = NumberFormatter()
        formatter.numberStyle = showsCurrencyCode ? .currencyISOCode : .currency
        formatter.currencyCode = currencyCode
        formatter.locale = Locale.autoupdatingCurrent
        let digits = Currency.fractionDigits(for: currencyCode)
        formatter.minimumFractionDigits = digits
        formatter.maximumFractionDigits = digits
        if Currency.symbol(for: currencyCode) != currencyCode && !showsCurrencyCode {
            formatter.currencySymbol = Currency.symbol(for: currencyCode)
        }
        storage[key] = formatter
        return formatter
    }
}
